### PR TITLE
Unify Label and DynLabel

### PR DESCRIPTION
**Pull request by zaynetro**
One annoying thing is that you have to specify that closure arguments are references. Otherwise, compiler will complain:

```diff
diff --git a/druid/examples/textbox.rs b/druid/examples/textbox.rs
index 7a5f918..55e0b65 100644
--- a/druid/examples/textbox.rs
+++ b/druid/examples/textbox.rs
@@ -44,7 +44,7 @@ fn build_widget() -> impl Widget<String> {
         },
         TextBox::new(),
     );
-    let label = Label::new(|data: &String, _env: &_| format!("value: {}", data));
+    let label = Label::new(|data: &String, _env| format!("value: {}", data));

     Flex::column()
         .with_child(Padding::new(5.0, textbox), 1.0)
```

Error:

```
error[E0271]: type mismatch resolving `for<'r, 's> <[closure@druid/examples/textbox.rs:47:28: 47:76] as std::ops::FnOnce<(&'r _, &'s druid::env::Env)>>::Output == std::string::String`
  --> druid/examples/textbox.rs:47:17
   |
47 |     let label = Label::new(|data: &String, _env| format!("value: {}", data));
   |                 ^^^^^^^^^^ expected bound lifetime parameter, found concrete lifetime
   |
   = note: required because of the requirements on the impl of `std::convert::From<[closure@druid/examples/textbox.rs:47:28: 47:76]>` for `druid::widget::label::LabelText<std::string::String>`
   = note: required because of the requirements on the impl of `std::convert::Into<druid::widget::label::LabelText<std::string::String>>` for `[closure@druid/examples/textbox.rs:47:28: 47:76]`
   = note: required by `druid::widget::label::Label::<T>::new`

error[E0631]: type mismatch in closure arguments
  --> druid/examples/textbox.rs:47:17
   |
47 |     let label = Label::new(|data: &String, _env| format!("value: {}", data));
   |                 ^^^^^^^^^^ ------------------------------------- found signature of `for<'r> fn(&'r std::string::String, _) -> _`
   |                 |
   |                 expected signature of `for<'r, 's> fn(&'r _, &'s druid::env::Env) -> _`
   |
   = note: required because of the requirements on the impl of `std::convert::From<[closure@druid/examples/textbox.rs:47:28: 47:76]>` for `druid::widget::label::LabelText<_>`
   = note: required because of the requirements on the impl of `std::convert::Into<druid::widget::label::LabelText<_>>` for `[closure@druid/examples/textbox.rs:47:28: 47:76]`
   = note: required by `druid::widget::label::Label::<T>::new`

error: aborting due to 2 previous errors
```

### CODE REVIEW

**Comment by cmyr** [comment]
sorry, been distracted the last couple days, but will give this a look over soon. :)

File: druid/src/widget/label.rs [hunk 1]
```diff
@@ -30,10 +28,15 @@ use crate::localization::LocalizedString;
...
+    /// Localized string that will be resolved through `Env`.
     Localized(LocalizedString<T>),
+    /// Specific text
     Specific(String),
+    /// The provided closure is called on update, and its return
+    /// value is used as the text for the label.
+    Dynamic(Box<dyn Fn(&T, &Env) -> String>),
 }
 
 /// A label that displays some text.
```
Comment (cmyr) [inline druid/src/widget/label.rs]: This would be a good time to plump up these docs a bit, showing the different ways the text can be provided?

**Comment by zaynetro** [comment]
Great idea about using the lens! Now that we have all those lens combinators it should be doable. I will experiment a bit.

**Comment by cmyr** [comment]
I think it's fine to merge this as is, and think of that work as future experimentation?

**Comment by zaynetro** [comment]
I didn't have time yet to take another look. Tomorrow is the day when I plan to do that. If nothing creative comes out then I will merge this as is.

**Comment by zaynetro** [comment]
I think I will leave it as is. Ready for merging once CI stops failing...

File: druid/src/widget/label.rs (hunk 1 above)
Comment (zaynetro) [reply druid/src/widget/label.rs]: Added usage docs.

**Comment by raphlinus** [comment]
This appears to be https://github.community/t5/GitHub-Actions/ubuntu-latest-Apt-repository-list-issues/td-p/41122 , and I don't see in that thread a clear resolution or workaround.

I'm personally ok with merging on red when the problem is clearly CI, but this also seems like something we need to fix.

**Comment by zaynetro** [comment]
Seems like the problem is fixed now. Thanks for finding the source!

**Pull request merged**